### PR TITLE
MM-18972

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
Migrated following javascript files to Typescript
`components/widgets/settings/`-
- [X] bool_setting & test
- [X] radio_setting & test
- [X] setting & test
- [X] text_setting & test

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12456

### CODE REVIEW

**Comment by M-ZubairAhmed** [comment]
@hanzei this is because we removed prop-types `missing in props validation react/prop-types` since we are using typescript

**Comment by M-ZubairAhmed** [comment]
Also let me know if this is a good approach (working with TS first time) so i can continue on other two files

File: yarn-error.log [hunk 1]
```diff
@@ -0,0 +1,261 @@
+Arguments:
```
Comment (devinbinnie) [inline yarn-error.log]: You shouldn't include this file in your PR.

File: components/widgets/settings/text_setting.tsx [hunk 2]
```diff
@@ -1,73 +1,65 @@
...
+    private handleChange: HandleChangeTypes = (e) => {
         if (this.props.type === 'number') {
             this.props.onChange(this.props.id, parseInt(e.target.value, 10));
         } else {
             this.props.onChange(this.props.id, e.target.value);
         }
     }
 
-    render() {
+    public render() {
```
Comment (devinbinnie) [inline components/widgets/settings/text_setting.tsx]: You can add the `JSX.Element` type on the `render()` functions

**Comment by M-ZubairAhmed** [comment]
> Hey @M-ZubairAhmed! Approach looks good, just a couple small comments.
> As for the lint errors, though should be disappearing once you remove `PropTypes` from the file, so I'm not sure what's happening there.

This seems to be happening only functional components not for class components

**Comment by M-ZubairAhmed** [comment]
@devinbinnie i have fixed the changes

File: components/widgets/settings/bool_setting.tsx [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
-        id: PropTypes.string.isRequired,
-        label: PropTypes.node.isRequired,
-        labelClassName: PropTypes.string,
-        helpText: PropTypes.node,
-        placeholder: PropTypes.string.isRequired,
-        value: PropTypes.bool.isRequired,
-        inputClassName: PropTypes.string,
-        onChange: PropTypes.func,
-    };
+type BoolSettingProps = {
```
Comment (saturninoabril) [inline components/widgets/settings/bool_setting.tsx]: nit: Please simply name it as `Props` (unless there are multiple components in a file).  Same case on other instance.

File: components/widgets/settings/bool_setting.tsx [hunk 4]
```diff
@@ -1,33 +1,32 @@
...
         inputClassName: '',
     };
 
-    handleChange = (e) => {
+    private handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
         this.props.onChange(this.props.id, e.target.checked);
     }
 
-    render() {
+    public render(): JSX.Element {
```
Comment (saturninoabril) [inline components/widgets/settings/bool_setting.tsx]: nit: I don't know if we came up with the decision on which to use for consistency but some (including me) preferred to make this inferred only (less noise) but @devinbinnie can confirm.
```suggestion
    public render() {
```

Same case on other instance.

File: components/widgets/settings/text_setting.tsx [hunk 5]
```diff
@@ -1,73 +1,65 @@
...
-            PropTypes.number,
-        ]).isRequired,
-        inputClassName: PropTypes.string,
-        maxLength: PropTypes.number,
-        resizable: PropTypes.bool,
-        onChange: PropTypes.func,
-        disabled: PropTypes.bool,
-        type: PropTypes.oneOf(TextSetting.validTypes),
-    };
+type InputTypes = 'input' |'textarea'| 'number'| 'email'| 'tel'| 'url'| 'password'
```
Comment (saturninoabril) [inline components/widgets/settings/text_setting.tsx]: Could you please fix spacing between types? (I wish linting can help with it but apparently not.)

File: components/widgets/settings/text_setting.tsx [hunk 6]
```diff
@@ -1,73 +1,65 @@
...
+type HandleChangeTypes = React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>
 
-    static defaultProps = {
+export default class TextSetting extends React.Component<TextSettingProps> {
+    public static defaultProps: Partial<TextSettingProps> = {
         labelClassName: '',
         inputClassName: '',
         type: 'input',
-        maxLength: null,
+        maxLength: -1, // A negative number allows for values of any lenght
```
Comment (saturninoabril) [inline components/widgets/settings/text_setting.tsx]:
```suggestion
        maxLength: -1, // A negative number allows for values of any length
```

File: components/widgets/settings/bool_setting.tsx (hunk 3 above)
Comment (M-ZubairAhmed) [reply components/widgets/settings/bool_setting.tsx]: @saturninoabril yes lint test are failing due to prop types not present although prop type is not being used in that file itself.
> This seems to be happening only functional components not for class components
What should be done for that ?

File: components/widgets/settings/bool_setting.tsx (hunk 4 above)
Comment (M-ZubairAhmed) [reply components/widgets/settings/bool_setting.tsx]: @saturninoabril i added that according to https://github.com/mattermost/mattermost-webapp/pull/4061#discussion_r339570026 from @devinbinnie

File: components/widgets/settings/text_setting.tsx (hunk 5 above)
Comment (M-ZubairAhmed) [reply components/widgets/settings/text_setting.tsx]: oh yes i will change this,
Also any specific reason for not running prettier, i am sure it could have formatted those.

File: components/widgets/settings/bool_setting.tsx (hunk 4 above)
Comment (devinbinnie) [reply components/widgets/settings/bool_setting.tsx]: @saturninoabril Yeah I'm not sure if there was any decisions made about that, I just figured it adds more clarity.

**Comment by devinbinnie** [comment]
@M-ZubairAhmed take a look at `components/loading_image_preview.tsx`, it's a functional component that is passing the linter. Maybe there's something in there that would help?

File: components/widgets/settings/bool_setting.tsx (hunk 4 above)
Comment (saturninoabril) [reply components/widgets/settings/bool_setting.tsx]: Thanks. I'm good with either approach, just a bit worried on consistency. Maybe should we start documenting style guide for Typescript, similar to what we have at https://docs.mattermost.com/developer/style-guide.html?

File: components/widgets/settings/text_setting.tsx (hunk 5 above)
Comment (saturninoabril) [reply components/widgets/settings/text_setting.tsx]: We looked into it before and there were format/styles that's different from what we set in our standard style.

**Comment by devinbinnie** [comment]
@M-ZubairAhmed You have some tests failing:

```
Summary of all failing tests
 FAIL  components/interactive_dialog/dialog_element/dialog_element.test.jsx
  ● components/interactive_dialog/DialogElement › subtype email

    TypeError: Cannot read property 'includes' of undefined

      127 |                 maxLength = maxLength || TEXT_DEFAULT_MAX_LENGTH;
      128 | 
    > 129 |                 if (subtype && TextSetting.validTypes.includes(subtype)) {
          |                                                       ^
      130 |                     type = subtype;
      131 |                 } else {
      132 |                     type = 'input';

      at DialogElement.includes [as render] (components/interactive_dialog/dialog_element/dialog_element.jsx:129:55)
      at ReactShallowRenderer._mountClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer-shallow.development.js:809:37)
      at ReactShallowRenderer.render (node_modules/react-test-renderer/cjs/react-test-renderer-shallow.development.js:738:14)
      at render (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:666:53)
      at fn (node_modules/enzyme-adapter-utils/src/Utils.js:99:18)
      at Object.render (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:666:18)
      at new render (node_modules/enzyme/src/ShallowWrapper.js:397:22)
      at shallow (node_modules/enzyme/src/shallow.js:10:10)
      at Object.it (components/interactive_dialog/dialog_element/dialog_element.test.jsx:50:25)

  ● components/interactive_dialog/DialogElement › subtype invalid

    TypeError: Cannot read property 'includes' of undefined

      127 |                 maxLength = maxLength || TEXT_DEFAULT_MAX_LENGTH;
      128 | 
    > 129 |                 if (subtype && TextSetting.validTypes.includes(subtype)) {
          |                                                       ^
      130 |                     type = subtype;
      131 |                 } else {
      132 |                     type = 'input';

      at DialogElement.includes [as render] (components/interactive_dialog/dialog_element/dialog_element.jsx:129:55)
      at ReactShallowRenderer._mountClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer-shallow.development.js:809:37)
      at ReactShallowRenderer.render (node_modules/react-test-renderer/cjs/react-test-renderer-shallow.development.js:738:14)
      at render (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:666:53)
      at fn (node_modules/enzyme-adapter-utils/src/Utils.js:99:18)
      at Object.render (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:666:18)
      at new render (node_modules/enzyme/src/ShallowWrapper.js:397:22)
      at shallow (node_modules/enzyme/src/shallow.js:10:10)
      at Object.it (components/interactive_dialog/dialog_element/dialog_element.test.jsx:65:25)

  ● components/interactive_dialog/DialogElement › subtype password

    TypeError: Cannot read property 'includes' of undefined

      127 |                 maxLength = maxLength || TEXT_DEFAULT_MAX_LENGTH;
      128 | 
    > 129 |                 if (subtype && TextSetting.validTypes.includes(subtype)) {
          |                                                       ^
      130 |                     type = subtype;
      131 |                 } else {
      132 |                     type = 'input';

      at DialogElement.includes [as render] (components/interactive_dialog/dialog_element/dialog_element.jsx:129:55)
      at ReactShallowRenderer._mountClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer-shallow.development.js:809:37)
      at ReactShallowRenderer.render (node_modules/react-test-renderer/cjs/react-test-renderer-shallow.development.js:738:14)
      at render (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:666:53)
      at fn (node_modules/enzyme-adapter-utils/src/Utils.js:99:18)
      at Object.render (node_modules/enzyme-adapter-react-16/src/ReactSixteenAdapter.js:666:18)
      at new render (node_modules/enzyme/src/ShallowWrapper.js:397:22)
      at shallow (node_modules/enzyme/src/shallow.js:10:10)
      at Object.it (components/interactive_dialog/dialog_element/dialog_element.test.jsx:80:25)

 FAIL  components/widgets/settings/text_setting.test.tsx
  ● components/widgets/settings/TextSetting › render component with required props

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `components/widgets/settings/TextSetting render component with required props 1`

    - Snapshot
    + Received

    @@ -5,11 +5,11 @@
        labelClassName=""
      >
        <input
          className="form-control"
          id="string.id"
    -     maxLength={null}
    +     maxLength={-1}
          onChange={[Function]}
          type="input"
          value="some value"
        />
      </Setting>

      17 |             />
      18 |         );
    > 19 |         expect(wrapper).toMatchInlineSnapshot(`
         |                         ^
      20 | <Setting
      21 |   inputClassName=""
      22 |   inputId="string.id"

      at Object.toMatchInlineSnapshot (components/widgets/settings/text_setting.test.tsx:19:25)

  ● components/widgets/settings/TextSetting › render with textarea type

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `components/widgets/settings/TextSetting render with textarea type 1`

    - Snapshot
    + Received

    @@ -5,12 +5,12 @@
        labelClassName=""
      >
        <textarea
          className="form-control"
          id="string.id"
    -     maxLength={null}
    +     maxLength={-1}
          onChange={[Function]}
    -     rows="5"
    +     rows={5}
          style={Object {}}
          value="some value"
        />
      </Setting>

      47 |             />
      48 |         );
    > 49 |         expect(wrapper).toMatchInlineSnapshot(`
         |                         ^
      50 | <Setting
      51 |   inputClassName=""
      52 |   inputId="string.id"

      at Object.toMatchInlineSnapshot (components/widgets/settings/text_setting.test.tsx:49:25)


Snapshot Summary
 › 2 snapshots failed from 1 test suite. Inspect your code changes or run `npm run test-ci -- -u` to update them.
```

**Comment by M-ZubairAhmed** [comment]
Working on it

**Comment by M-ZubairAhmed** [comment]
@devinbinnie thanks for the suggestion , all tests are now passing, please review the the pr

File: package-lock.json [hunk 7]
```diff
@@ -3352,6 +3352,16 @@
       "resolved": "https://registry.npmjs.org/@types/history/-/history-4.7.3.tgz",
       "integrity": "sha512-cS5owqtwzLN5kY+l+KgKdRJ/Cee8tlmQoGQuIE9tWnSmS3JMKzmxo2HIAk2wODMifGwO20d62xZQLYz+RLfXmw=="
     },
+    "@types/hoist-non-react-statics": {
```
Comment (saturninoabril) [inline package-lock.json]: All looks good to me except to revert this change since no change in `package.json`.

**Comment by M-ZubairAhmed** [comment]
How is this state PR, when it was approved by the team member ??

**Comment by jasonblais** [comment]
Hey @M-ZubairAhmed! Looks like it just needs one more review from @catalintomai and it's ready to go. 

@catalintomai when you have time, can you help review this PR from @M-ZubairAhmed?

**Comment by catalintomai** [comment]
> Hey @M-ZubairAhmed! Looks like it just needs one more review from @catalintomai and it's ready to go.
> 
> @catalintomai when you have time, can you help review this PR from @M-ZubairAhmed?

@jasonblais , how many core committer review approvals do community members PRs need? I believe this one had two already.

**Comment by jasonblais** [comment]
@catalintomai Each PR requires two dev reviews, and a QA review where applicable. I believe Saturn's review would apply for QA review in this case.

**Comment by catalintomai** [comment]
> @catalintomai Each PR requires two dev reviews, and a QA review where applicable. I believe Saturn's review would apply for QA review in this case.

Thanks. Saturnino's infotag above has "Core Committer" in it. Maybe we should update the [doc](https://developers.mattermost.com/contribute/getting-started/code-review/) and explicitly refer to **dev** core committers as opposed to QA Core Committers/or QA short.

**Comment by M-ZubairAhmed** [comment]
I am very impressed by review process of this project, let me check out more issues so i can contribute to this project.
